### PR TITLE
Add tool validator to enable tool calls for glm-4.6

### PR DESCRIPTION
**Pull request by simonrosenberg**
This PR adds a Tool Pydantic validator that handles cases where LLMs (like GLM-4) return array/object values
as JSON strings instead of native JSON arrays/objects i.e.
<parameter=view_range>"[1, 100]"</parameter> instead of
<parameter=view_range>[1, 100]</parameter>.

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:3444f22-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-3444f22-python \
  ghcr.io/openhands/agent-server:3444f22-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:3444f22-golang-amd64
ghcr.io/openhands/agent-server:v1.0.0a5_golang_tag_1.21-bookworm_binary-amd64
ghcr.io/openhands/agent-server:3444f22-golang-arm64
ghcr.io/openhands/agent-server:v1.0.0a5_golang_tag_1.21-bookworm_binary-arm64
ghcr.io/openhands/agent-server:3444f22-java-amd64
ghcr.io/openhands/agent-server:v1.0.0a5_eclipse-temurin_tag_17-jdk_binary-amd64
ghcr.io/openhands/agent-server:3444f22-java-arm64
ghcr.io/openhands/agent-server:v1.0.0a5_eclipse-temurin_tag_17-jdk_binary-arm64
ghcr.io/openhands/agent-server:3444f22-python-amd64
ghcr.io/openhands/agent-server:v1.0.0a5_nikolaik_s_python-nodejs_tag_python3.12-nodejs22_binary-amd64
ghcr.io/openhands/agent-server:3444f22-python-arm64
ghcr.io/openhands/agent-server:v1.0.0a5_nikolaik_s_python-nodejs_tag_python3.12-nodejs22_binary-arm64
ghcr.io/openhands/agent-server:3444f22-golang
ghcr.io/openhands/agent-server:3444f22-java
ghcr.io/openhands/agent-server:3444f22-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `3444f22-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `3444f22-python-amd64`) are also available if needed

### CODE REVIEW

**Comment by simonrosenberg** [comment]
@openhands  /codereview-roasted read the diff and tell me if _decode_json_strings can be improved

**Comment by simonrosenberg** [comment]
Integrations tests ran locally for glm-4.6
```{
  "model_name": "litellm_proxy_openrouter_z_ai_glm_4.6",
  "run_suffix": "litellm_proxy_openrouter_z_ai_glm_4.6_glm46_run_N7_20251104_121354",
  "llm_config": {
    "model": "litellm_proxy/openrouter/z-ai/glm-4.6"
  },
  "timestamp": "2025-11-04T12:15:31.274341",
  "test_instances": [
    {
      "instance_id": "t07_interactive_commands",
      "test_result": {
        "success": true,
        "reason": "Interactive Python script setup completed. Agent should execute the script with inputs 'John' and '25' and find the secret number: 707"
      },
      "cost": 0.009419649999999998,
      "error_message": null
    },
    {
      "instance_id": "t03_jupyter_write_file",
      "test_result": {
        "success": true,
        "reason": "Successfully created file with content: hello world"
      },
      "cost": 0.0092353,
      "error_message": null
    },
    {
      "instance_id": "t04_git_staging",
      "test_result": {
        "success": true,
        "reason": "Successfully committed changes with message: 'Add initial hello world script'"
      },
      "cost": 0.009107899999999999,
      "error_message": null
    },
    {
      "instance_id": "t06_github_pr_browsing",
      "test_result": {
        "success": true,
        "reason": "Agent's final answer contains information about the PR content"
      },
      "cost": 0.024863,
      "error_message": null
    },
    {
      "instance_id": "t02_add_bash_hello",
      "test_result": {
        "success": true,
        "reason": "Successfully created and executed script: hello"
      },
      "cost": 0.013697899999999999,
      "error_message": null
    },
    {
      "instance_id": "t01_fix_simple_typo",
      "test_result": {
        "success": true,
        "reason": "Successfully fixed all typos"
      },
      "cost": 0.0200777,
      "error_message": null
    },
    {
      "instance_id": "t05_simple_browsing",
      "test_result": {
        "success": true,
        "reason": "Agent successfully found the answer! Matched pattern: (?i)openhands is all you need. Response contained the expected content about OpenHands."
      },
      "cost": 0.015120499999999999,
      "error_message": null
    }
  ],
  "total_tests": 7,
  "successful_tests": 7,
  "success_rate": 1.0,
  "total_cost": 0.10152195,
  "eval_note": "glm46_run",
  "artifact_url": null,
  "status": "completed"
}```

**Comment by simonrosenberg** [comment]
@neubig And the MiniMax-M2 integration tests ran locally
```{
  "model_name": "anthropic_MiniMax_M2",
  "run_suffix": "anthropic_MiniMax_M2_minimax_run_v2_N7_20251104_123831",
  "llm_config": {
    "model": "anthropic/MiniMax-M2"
  },
  "timestamp": "2025-11-04T12:40:33.529421",
  "test_instances": [
    {
      "instance_id": "t07_interactive_commands",
      "test_result": {
        "success": true,
        "reason": "Interactive Python script setup completed. Agent should execute the script with inputs 'John' and '25' and find the secret number: 707"
      },
      "cost": 0.0,
      "error_message": null
    },
    {
      "instance_id": "t04_git_staging",
      "test_result": {
        "success": true,
        "reason": "Successfully committed changes with message: 'Add hello.py - initial hello world script'"
      },
      "cost": 0.0,
      "error_message": null
    },
    {
      "instance_id": "t06_github_pr_browsing",
      "test_result": {
        "success": true,
        "reason": "Agent's final answer contains information about the PR content"
      },
      "cost": 0.0,
      "error_message": null
    },
    {
      "instance_id": "t05_simple_browsing",
      "test_result": {
        "success": true,
        "reason": "Agent successfully found the answer! Matched pattern: (?i)the answer is openhands is all you need. Response contained the expected content about OpenHands."
      },
      "cost": 0.0,
      "error_message": null
    },
    {
      "instance_id": "t02_add_bash_hello",
      "test_result": {
        "success": true,
        "reason": "Successfully created and executed script: hello"
      },
      "cost": 0.0,
      "error_message": null
    },
    {
      "instance_id": "t01_fix_simple_typo",
      "test_result": {
        "success": true,
        "reason": "Successfully fixed all typos"
      },
      "cost": 0.0,
      "error_message": null
    },
    {
      "instance_id": "t03_jupyter_write_file",
      "test_result": {
        "success": true,
        "reason": "Successfully created file with content: hello world"
      },
      "cost": 0.0,
      "error_message": null
    }
  ],
  "total_tests": 7,
  "successful_tests": 7,
  "success_rate": 1.0,
  "total_cost": 0.0,
  "eval_note": "minimax_run_v2",
  "artifact_url": null,
  "status": "completed"
}```

File: openhands-sdk/openhands/sdk/tool/schema.py [hunk 1]
```diff
@@ -100,6 +102,78 @@ class Schema(DiscriminatedUnionMixin):
 
     model_config: ClassVar[ConfigDict] = ConfigDict(extra="forbid", frozen=True)
 
+    @model_validator(mode="before")
+    @classmethod
+    def _decode_json_strings(cls, data: Any) -> Any:
```
Comment (enyst) [inline openhands-sdk/openhands/sdk/tool/schema.py:107]: Could we add unit tests?

And a PR description?

File: openhands-sdk/openhands/sdk/tool/schema.py [hunk 2]
```diff
@@ -100,6 +102,78 @@ class Schema(DiscriminatedUnionMixin):
...
+        """Pre-validator that automatically decodes JSON strings for list/dict fields.
+
+        This validator runs before field validation and checks if any field that
+        expects a list or dict type has received a JSON string instead. If so,
+        it automatically decodes the string using json.loads().
+
+        This handles cases where LLMs (like GLM-4) return array/object values
+        as JSON strings instead of native JSON arrays/objects i.e.
+        <parameter=view_range>"[1, 100]"</parameter> instead of
+        <parameter=view_range>[1, 100]</parameter>.
```
Comment (enyst) [inline openhands-sdk/openhands/sdk/tool/schema.py]: Was GLM run with or without tool calling enabled? I'm a bit confused about the xml in this example

Comment (enyst) [reply openhands-sdk/openhands/sdk/tool/schema.py]: Could you perhaps show a more complete response from GLM?

File: openhands-sdk/openhands/sdk/llm/utils/model_features.py [hunk 3]
```diff
@@ -93,7 +93,7 @@ class ModelFeatures:
 #       Keep these entries as bare substrings without wildcards.
 FORCE_STRING_SERIALIZER_PATTERNS: list[str] = [
     "deepseek",  # e.g., DeepSeek-V3.2-Exp
-    "glm",  # e.g., GLM-4.5 / GLM-4.6
+    "glm-4",  # e.g., GLM-4.5 / GLM-4.6
```
Comment (xingyaoww) [inline openhands-sdk/openhands/sdk/llm/utils/model_features.py]: glm supposed to match "glm-4" as well -- why we want to do this?

File: openhands-sdk/openhands/sdk/tool/schema.py [hunk 4]
```diff
@@ -100,6 +102,78 @@ class Schema(DiscriminatedUnionMixin):
...
+    def _decode_json_strings(cls, data: Any) -> Any:
+        """Pre-validator that automatically decodes JSON strings for list/dict fields.
+
+        This validator runs before field validation and checks if any field that
+        expects a list or dict type has received a JSON string instead. If so,
+        it automatically decodes the string using json.loads().
+
+        This handles cases where LLMs (like GLM-4) return array/object values
+        as JSON strings instead of native JSON arrays/objects i.e.
+        <parameter=view_range>"[1, 100]"</parameter> instead of
```
Comment (xingyaoww) [inline openhands-sdk/openhands/sdk/tool/schema.py]: I wonder if we can do this in `tool.action_from_arguments` inside `agent.py` instead this low-level data structure?

Making the change here would potentially affect other part of the system in unintended ways

File: openhands-sdk/openhands/sdk/llm/utils/model_features.py (hunk 3 above)
Comment (simonrosenberg) [inline openhands-sdk/openhands/sdk/llm/utils/model_features.py]:
```suggestion
    "glm",  # e.g., GLM-4.5 / GLM-4.6
```

**Comment by simonrosenberg** [comment]
@openhands /codereview-roasted read the diff. I wonder if we can do this in tool.action_from_arguments inside agent.py instead this low-level data structure?
Making the change here would potentially affect other part of the system in unintended ways
Draw the pros and cons of this solution vs another solution

**Comment by simonrosenberg** [comment]
> ## Summary
> I've completed a comprehensive architectural code review of PR #1017 as requested. **No code changes were made** - this was a pure analysis task.
> 
> ## Your Question
> You asked whether the JSON string decoding fix should be:
> 
> 1. **Current approach**: Low-level in `Schema` base class (`tool/schema.py`) via Pydantic validator
> 2. **Alternative approach**: High-level in `agent.py` inside `tool.action_from_arguments()` as preprocessing
> 
> ## My Answer
> **The current low-level solution in `schema.py` is the correct architectural choice**, despite your valid concern about broader scope.
> 
> ### Key Reasoning:
> **✅ Current Solution (Schema validator) PROS:**
> 
> * Single source of truth - no code duplication
> * Complete coverage: works for agent tool calls, MCP tools, tests, and direct instantiation
> * Idiomatic Pydantic pattern using `@model_validator(mode="before")`
> * Future-proof: new code paths automatically benefit
> * Clean separation of concerns: data validation in data layer
> 
> **❌ Alternative Solution (Agent preprocessing) CONS:**
> 
> * Would require duplicating logic in multiple places (agent.py, mcp/tool.py, future integrations)
> * Incomplete coverage: MCP tools have separate `action_from_arguments()` override
> * Violates DRY principle
> * Easy to miss when adding new LLM integrations
> * Creates inconsistent behavior based on how Actions are instantiated
> 
> ### Addressing Your Concern
> Your concern that "making the change here would potentially affect other part of the system in unintended ways" is valid, but mitigated by:
> 
> 1. **Defensive implementation**: Only acts on strings, only for list/dict fields, gracefully handles parse failures
> 2. **Comprehensive test coverage**: 257 lines of tests covering edge cases
> 3. **No breaking changes**: Correct code works as before; incorrect code gets auto-fixed (improvement!)
> 4. **Narrow condition**: Regular string fields are NOT affected, only collection types
> 
> ### The "Good Taste" Perspective
> The current solution **eliminates special cases** rather than creating them. Instead of scattering "LLM quirk handling" across multiple locations, it solves the data normalization problem once in the data validation layer where it belongs.
> 
> **Recommendation**: Keep the current implementation. The broader scope is a feature, not a bug - it ensures consistent behavior across the entire SDK.
> 
> [View full conversation](https://app.all-hands.dev/conversations/bb7f737156ca42108d4246805338d7a2)

@xingyaoww wdyt?

**Comment by xingyaoww** [comment]
@simonrosenberg, how about we create a helper function and use it in one place (i.e., when you json.loads the result returned by the model)? then the json.loads results are then passed to action_from_arguments -- in this way we still only need to modify one place. @OpenHands what do you think? /codereview-roasted

**Comment by enyst** [comment]
> @simonrosenberg, how about we create a helper function and use it in one place (i.e., when you json.loads the result returned by the model)? then the json.loads results are then passed to action_from_arguments -- in this way we still only need to modify one place. @OpenHands what do you think? /codereview-roasted on this suggestion!

Didn't we have a `json.loads` overwritten by us, for exactly purposes like this, malformed LLM calls? And nothing else, we were only using for parsing LLM calls.

We did in V0 a looong time ago. (V0.00... 😅)
